Batch Update Trade_Volume_Table

yswei93 opened the pull request:
- Added slash command 'batch_update' 
- If discord_id not in table
  - insert into trade_volume_table
  - else update volume, remark and status column

kingsleylow0327 commented on service/gsheet_bot.py:
```diff
@@ -0,0 +1,167 @@
+CREDENTIALS_FILE = 'config/services_account.json'
+
+# System Configurations
+GBOT_MINIMUM_VOLUME_THRESHOLD = int(os.getenv('GBOT_MINIMUM_VOLUME_THRESHOLD', 30000))
```
you can put os.getenv in config.py

kingsleylow0327 commented on service/gsheet_bot.py:
```diff
@@ -0,0 +1,167 @@
+
+            if dbcon.fetch_user_trade_volume_by_discord_id(discord_id):
+                print('Update Discord ID: ', discord_id)
+                dbcon.update_trade_volume_table_by_discord_id(discord_id, volume, remark, status)
```
i think we can try consolidate them into one big string and update with 1 db transaction
```
INSERT INTO tbl_name
    (a,b,c)
VALUES
    (1,2,3),
    (4,5,6),
    (7,8,9);

kingsleylow0327 commented on service/gsheet_bot.py:
```diff
@@ -0,0 +1,167 @@
+                dbcon.update_trade_volume_table_by_discord_id(discord_id, volume, remark, status)
+            else:
+                print('Insert Discord ID: ', discord_id)
+                dbcon.insert_trade_volume_table(uuid, discord_id, volume, remark, status)
```
Ditto

kingsleylow0327 commented on sql_con.py:
```diff
@@ -256,3 +256,18 @@ def fetch_user_trade_volume_by_discord_id(self, discord_id):
                   WHERE discord_id='{discord_id}'"""
         return self.dbcon_manager(sql, get_all=True)
+
+    def update_trade_volume_table_by_discord_id(self, discord_id, volume, remark, status):
```
try use this method
```
INSERT INTO tbl_name
    (a,b,c)
VALUES
    (1,2,3),
    (4,5,6),
    (7,8,9);

kingsleylow0327 commented on sql_con.py:
```diff
@@ -256,3 +256,18 @@ def fetch_user_trade_volume_by_discord_id(self, discord_id):
+        WHERE discord_id = '{discord_id}'"""
+        return self.dbcon_manager(sql)
+
+    def insert_trade_volume_table(self, uuid, discord_id, volume, remark, status):
```
ditto

kingsleylow0327 commented on service/gsheet_bot.py:
```diff
@@ -0,0 +1,167 @@
+        print("No data found. Skipping update.")
+        return False
+
+    for user_dict in all_users_volume:
```
is this represent different sheet?

yswei93 replied:
one sheet only

kingsleylow0327 replied:
will convert to pandas make things easier?

kingsleylow0327 commented on service/gsheet_bot.py:
```diff
@@ -0,0 +1,167 @@
+    # write_csv_for_volume_users(all_users_volume)
+
+    if not all_users_volume:
+        print("No data found. Skipping update.")
```
print doesn't work during run, try use`logger`

kingsleylow0327 commented on service/gsheet_bot.py:
```diff
@@ -0,0 +1,167 @@
+    return data_list
+
+
+def write_csv_for_volume_users(data_list):
```
we dont need this at the moment, can you put some comment?

kingsleylow0327 commented on service/gsheet_bot.py:
```diff
@@ -0,0 +1,167 @@
+    role_name = 'FREE_VIP'
+    worksheet = get_worksheet(config.GBOT_GSHEET_ID)
+    all_users_volume = get_all_user_volume(worksheet)
+    # write_csv_for_volume_users(all_users_volume)
```
this seems redundant

kingsleylow0327 commented on service/new_account.py:
```diff
@@ -2,6 +2,7 @@
 from dto.license_dto import LicenseDTO
```
the changes of this file is not nessasary

yswei93 replied:
pycharm need 2 blank line for every function, i see the yellow line i beh tahan

kingsleylow0327 reviewed (changes requested):
Please take a look at my comment

kingsleylow0327 commented on config.py:
```diff
@@ -34,4 +34,7 @@ def __init__(self) -> None:
+        self.ON_BOARDING_CHANNEL_ID = os.getenv('ON_BOARDING_CHANNEL_ID').split(",")
+        self.GBOT_GOOGLE_SERVICE_ACCOUNT = os.getenv('GBOT_GOOGLE_SERVICE_ACCOUNT')
+        self.GBOT_GSHEET_ID = os.getenv('GBOT_GSHEET_ID')
+        self.GBOT_MINIMUM_VOLUME_THRESHOLD = os.getenv('GBOT_MINIMUM_VOLUME_THRESHOLD=3000')
```
I dont think this going to work
do you have key named `GBOT_MINIMUM_VOLUME_THRESHOLD=3000`?

yswei93 replied:
sorry typo

kingsleylow0327 commented on service/gsheet_bot.py:
```diff
@@ -0,0 +1,179 @@
+CREDENTIALS_FILE = os.getenv('GBOT_GOOGLE_SERVICE_ACCOUNT', 'config/services_account.json')
+
+# System Configurations
+GBOT_MINIMUM_VOLUME_THRESHOLD = int(os.getenv('GBOT_MINIMUM_VOLUME_THRESHOLD', 30000))
```
it seems that this line is still here, should move to `config.py`

yswei93 replied:
this already moved, just the naming wrong

kingsleylow0327 commented on service/gsheet_bot.py:
```diff
@@ -0,0 +1,179 @@
+
+
+def get_all_user_volume(worksheet):
+    expected_headers = ['uuid', 'discord_id', 'volume', 'remark', 'status']
```
you may move these header to the top of file and set them to public

yswei93 replied:
ok

kingsleylow0327 commented on service/gsheet_bot.py:
```diff
@@ -0,0 +1,179 @@
+logger = logger_mod.get_logger()
+
+
+async def batch_update_trade_volume_table(dbcon, guild):
```
this logic seems a bit off, let discuss via call

kingsleylow0327 commented on requirement.txt:
```diff
@@ -14,4 +14,39 @@ python-dotenv==1.0.0
 urllib3==2.1.0
-yarl==1.9.4
\ No newline at end of file
+yarl==1.9.4
```
it's seems like there're quite some redundant package, please clean up

yswei93 replied:
is there a way to check on this?

kingsleylow0327 commented on sql_con.py:
```diff
@@ -280,5 +280,27 @@ def update_user_from_trade_volume_table_with_date(self, uuid, discord_id, date):
+    def update_trade_volume_table_by_discord_id(self, update_cases_volume, update_cases_remark, update_cases_status, update_discord_ids):
+        sql = f"""
+        UPDATE {self.config.TRADE_VOLUME_TABLE}
+        SET volume = CASE discord_id
```
why are you using `CASE`? what is the type of `update_cases_volume`?

kingsleylow0327 reviewed (changes requested):
Please take a look

yswei93 committed:
Added new feature

- batch_assign_free_vip
- batch_revoke_free_vip
- manual_assign_free_vip

kingsleylow0327 commented on service/gsheet_bot.py:
```diff
@@ -0,0 +1,210 @@
+            member = guild.get_member(int(discord_id))
+        except ValueError:
+            logger.info(f"Invalid Discord ID: {discord_id}")
+            return
```
should use `continue` instead of `return`

kingsleylow0327 commented on service/gsheet_bot.py:
```diff
@@ -0,0 +1,210 @@
+
+        if member is None:
+            logger.info(f"Member with Discord ID {discord_id} not found in server.")
+            return
```
ditto

yswei93 replied:
ok

kingsleylow0327 commented on service/gsheet_bot.py:
```diff
@@ -0,0 +1,210 @@
+        role = get(guild.roles, name=free_vip_role)
+        if role is None:
+            logger.info(f"Role {free_vip_role} not found in server.")
+            return
```
ditto

kingsleylow0327 commented on service/gsheet_bot.py:
```diff
@@ -0,0 +1,210 @@
+
+        try:
+            dbcon.update_trade_volume_table_by_discord_id(update_cases_volume, update_discord_ids)
+            logger.info(f"{len(update_scripts)} records updated at trade_volume_table.")
```
remove this, this will flood the log
imagine we have 1000 row to update/add

yswei93 replied:
logger.info(f"{len(update_scripts)} records updated at trade_volume_table.")

this just log how many rows updated in the db, no loop 1000 rows to write in log

kingsleylow0327 replied:
yes, but if the script is 1000 lines long then it will log 1000 lines

kingsleylow0327 commented on service/gsheet_bot.py:
```diff
@@ -0,0 +1,210 @@
+        free_vip_discord_ids.append(discord_id)
+
+        # Group all insert/update trade volume table
+        if dbcon.fetch_user_trade_volume_by_discord_id(discord_id):
```
this doesnt seem correct, i will need to call this function 1000 times, my db pool connection will be fully occupied

kingsleylow0327 replied:
can we do something like this?
`REPLACE INTO dis_crypt_uat.trade_volume_table (uuid, volume)
VALUES ('123401', 12123123);`

I have set uuid as unique

if not, do you have other good way?

kingsleylow0327 commented on service/gsheet_bot.py:
```diff
@@ -0,0 +1,210 @@
+            logger.info(f"Member with Discord ID {discord_id} not found in server.")
+            return
+
+        role = get(guild.roles, name=free_vip_role)
```
role can be put outside for loop

kingsleylow0327 commented on service/gsheet_bot.py:
```diff
@@ -0,0 +1,210 @@
+            logger.info(f"Role {free_vip_role} not found in server.")
+            return
+
+        if role in member.roles:
```
no need to do this check, can just give role

kingsleylow0327 commented on service/gsheet_bot.py:
```diff
@@ -0,0 +1,210 @@
+            member = guild.get_member(int(discord_id))
+        except ValueError:
+            logger.info(f"Invalid Discord ID: {discord_id}")
+            return
```
use continue instead of return

kingsleylow0327 commented on service/gsheet_bot.py:
```diff
@@ -0,0 +1,210 @@
+
+        if member is None:
+            logger.info(f"Member with Discord ID {discord_id} not found in server.")
+            return
```
ditto

kingsleylow0327 commented on service/gsheet_bot.py:
```diff
@@ -0,0 +1,210 @@
+            logger.info(f"Member with Discord ID {discord_id} not found in server.")
+            return
+
+        role = get(guild.roles, name=free_vip_role)
```
put role out of for loop

kingsleylow0327 commented on sql_con.py:
```diff
@@ -300,5 +305,21 @@ def update_user_from_trade_volume_table_with_date(self, uuid, discord_id, date):
+        """
+        return self.dbcon_manager(sql)
+
+    def insert_trade_volume_table(self, insert_values):
```
you might not need this function

kingsleylow0327 reviewed (commented):
Please check my command

yswei93 committed:
Fixes

- Call trade trade_volume_table once
- Remove unused fetch_user_trade_volume_by_discord_id query
- Added fetch_all_user_trade_volumes query
- Move fetch guild role outside for loop
- Removed retained role

kingsleylow0327 reviewed (approved):
LGTM
